### PR TITLE
fix(install): fix the case that package scope contains tilde.

### LINKED ISSUE #7045: bun can't install  @~3/svelte_mount in cli
### What version of Bun is running?

1.0.11

### What platform is your computer?

Darwin 23.1.0 arm64 arm

### What steps can reproduce the bug?

bun i @~3/svelte_mount

```
bun add v1.0.11 (f7f6233e)
  🔍 Resolving [1/1]
error: InstallFailed cloning repository for @~3/svelte_mou

error: "git clone" for "@~3/svelte_mount" failed


error: "git clone" for "@~3/svelte_mount" failed
````

### What is the expected behavior?

image

### What do you see instead?

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Hanaasagi**
### What does this PR do?

Close #7045 

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

I wrote automated tests

### CODE REVIEW

**Comment by Jarred-Sumner** [comment]
huh that is expected to work? interesting

**Review by Jarred-Sumner — APPROVED** [review]

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Actually, this code should check that the path name between `@` and `/` is a valid URL component. 

Something like this:
```zig
pub fn hasURLComponentThatNeedsEncoding(str: []const u8) bool {
        const do_not_escape_when_encoding_uri_component = comptime brk: {
            // encodeURIComponent uses this table:
            const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ" ++
                "abcdefghijklmnopqrstuvwxyz" ++
                "0123456789" ++
                "!'()*-._~";

            var bitset = std.bit_set.IntegerBitSet(std.math.maxInt(u8)).initFull();
            for (chars) |char| {
                bitset.unset(@as(u8, char));
            }

            break :brk bitset;
        };

        for (str) |char| {
            if (do_not_escape_when_encoding_uri_component.isSet(char)) {
                return true;
            }
        }

        return false;
    }
```

npm uses `encodeURIComponent`

File: src/string_immutable.zig [hunk 1]
```diff
@@ -144,16 +144,26 @@ pub inline fn isNPMPackageName(target: string) bool {
         '@' => true,
         else => return false,
     };
+
     var slash_index: usize = 0;
     for (target[1..], 0..) |c, i| {
         switch (c) {
             // Old packages may have capital letters
-            'A'...'Z', 'a'...'z', '0'...'9', '$', '-', '_', '.' => {},
+            'A'...'Z', 'a'...'z', '0'...'9', '-', '_', '.' => {},
```
Comment (Hanaasagi) [inline src/string_immutable.zig:152]: Should I rewrite these switch branches to bitmap?

File: src/string_immutable.zig [hunk 2]
```diff
@@ -144,16 +144,26 @@ pub inline fn isNPMPackageName(target: string) bool {
...
+            '$' => {
```
Comment (dylan-conway) [inline src/string_immutable.zig]: is `$` valid? It should be escaped by `encodeURIComponent` so we can remove the case for it

Comment (Hanaasagi) [reply src/string_immutable.zig]: 🤯 sorry, fixed.

**Review by dylan-conway — APPROVED** [review]

**Comment by dylan-conway** [comment]
Thank you!

**Pull request merged**